### ISSUE TITLE
iPad Pro 3rd Generation - Framing for app store

### ISSUE THREAD

**Issue by Alarson93** [fastlane/fastlane#14713]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

(Well technically I'm using 2.121.1 - but I suspect the minor version change hasn't addressed this based on the frame files I see on Github.)

### Issue Description

I cannot frame iPad Pro 3rd Gen screenshots. Framing other devices (XS Max, 8 Plus, etc) works just fine.

I assume that this is a result of Facebook not having created iPad Pro (3rd gen) device frames. If this is the case - our designer has created frames that you could perhaps use. 

Screenshots work fine - I can even upload unframed screenshots of the Pro 3rd Gen - this is strictly a framing issue. 

##### Command executed

```
  lane :take_screenshots do
    # Flavor 
    setup_for_external_demo

    # Snapshot
    devices = ["iPad Pro (12.9-inch) (3rd generation)"] # Only do what is required by the store ### ["iPhone 8 Plus", "iPhone Xs Max", "iPad Pro (12.9-inch)", "iPad Pro (12.9-inch) (3rd generation)"]
    capture_screenshots(scheme:"Foobar", stop_after_first_error: true, reinstall_app: true, number_of_retries: 0, skip_helper_version_check: true, devices: devices, output_directory: ENV["SNAPSHOT_DIRECTORY"])

    # Frame
    makeBackgroundTileCommand = "convert -size 1x1 xc:#{ENV['PRIMARY_COLOR']} #{ENV['WORKSPACE']}/#{ENV['SNAPSHOT_DIRECTORY']}/Background.jpg"
    puts("Calling: ")
    puts(makeBackgroundTileCommand)

    makeBackgroundTileCommandOutput = `#{makeBackgroundTileCommand}`
    puts(makeBackgroundTileCommandOutput)

    frameit(path: ENV["SNAPSHOT_DIRECTORY"])

    # Upload
    upload_to_app_store(force: true, overwrite_screenshots: true)
  end
```
##### Complete output when running fastlane, including the stack trace and command used

I'm only including a relevant sample.
<details>
  <pre>[16:34:34]: ---------------------
[16:34:34]: --- Step: frameit ---
[16:34:34]: ---------------------
[16:34:34]: Framing screenshots at path Screenshots (via frameit)
[16:34:34]: Using device frames version 'latest'
[16:34:34]: Unknown device type for size iOS-iPad-Pro-12.9 for path './en-US/iPad Pro (12.9-inch) (3rd generation)-7foo1234.png'
[16:34:34]: Framing screenshot './en-US/iPad Pro (12.9-inch) (3rd generation)-7foo1234.png'
[16:34:34]: Unknown device type for size iOS-iPad-Pro-12.9 for path './en-US/iPad Pro (12.9-inch) (3rd generation)-7foo1234.png'
[16:34:34]: Unknown device type for size iOS-iPad-Pro-12.9 for path './en-US/iPad Pro (12.9-inch) (3rd generation)-7foo1234.png'
[16:34:34]: Couldn't find template for screenshot type 'Apple true Silver'
[16:34:34]: Please run `fastlane frameit download_frames` to download the latest frames
[16:34:34]: Unknown device type for size iOS-iPad-Pro-12.9 for path './en-US/iPad Pro (12.9-inch) (3rd generation)-10bar1234.png'</pre>
</details>

### Environment

I define my strings via `title.strings`.

Framefile:
<details>
  <pre>{
	"default": {
		"title": {
			"color": "#FFFFFFFF"
		},
    	"background": "Background.jpg",
		"padding": 50,
		"frame": "WHITE"
	},
	"data": []
}</pre>
</details>

**Comment by janpio** [comment]
Yes, the main problem is that we don't have those frames with a license that allows us to use them for this. Can you provide that and publish them e.g. on a GitHub repository?

**Comment by Alarson93** [comment]
I certainly can - I've already received a thumbs up from our designer.

@janpio How would you suggest I publish them? (link to a repo the contains it, pull request, post the frame on this thread, etc)

**Comment by janpio** [comment]
The frames in https://github.com/fastlane/frameit-frames, which are downloaded and used by frameit, are created by a script here: https://github.com/fastlane/fastlane/tree/master/frameit/frames_generator

Looking at https://github.com/fastlane/fastlane/blob/master/frameit/frames_generator/Rakefile the simplest would probably be to create a GitHub repository, put the files there in a folder structure that matches the Facebook Design one (`Tablets\Apple iPad Pro XXX\Device`, where `XXX` is replaced with how we decide to name the screenshots). Then we can use the GitHub Download repo feature to download a zip that can be extracted and put into the same `raw` folder as the Facebook frames.

The script then takes those files, removed the white padding around them and measures the inner offset and writes them to a file - but that should probably work with your images as well, and if not we can still fix that later on.

@mollyIV You worked on this recently, do you think my suggestion would work?
@liamnichols What what be an appropriate file name so this works with frameit best?

**Comment by liamnichols** [comment]
This is gonna be a bit tricky again 😅
The current strategy is `Apple {device_name} {frame_colour}.png` so I guess that it should be `Apple iPad Pro 12.9 inch (3rd Generation) Space Gray.png`

It seems a bit over the top, but I think that’s how it has to be

**Comment by janpio** [comment]
So be it. @Alarson93 can you work with that information? When you have that all in a public GitHub repo post its URL here and we can take over from there and see if it works (and iterate if necessary).

**Comment by mollyIV** [comment]
> @mollyIV You worked on this recently, do you think my suggestion would work?

@janpio I think your suggestion would work, as long as the catalog structure and naming convention are the same as the Facebook Design one 👍 Feel free to ping me when the GitHub repository is ready, I will be happy to make the changes in the frameit tool 🙇

**Comment by Alarson93** [comment]
(Sorry for the delay - I was very busy yesterday!)

 Thank you for those instructions. They should suffice. If I have any questions I can always ask!

That said - I appreciate you mentioning licensing. I asked our designer about that and she said that she had download them from somewhere, but doesn't remember where. I was under the impression she made them.
She is going to check to see the usage conditions when she has time. Worst case scenario - I'm sure we could make our own. 

I have the feeling that this will probably drag on through the week, but I will make sure that I update you as things develop and let you know when a repo is set up.

I appreciate all of your support!

**Comment by Alarson93** [comment]
Update: 
Our designer said she got it from [here](https://applypixels.com/resource/ipad-pro-3rd-gen).

Here is the [license](https://applypixels.com/page/license). 
It sounds pretty open, but I think the sticking point would be on re-distrubtion. 

> You cannot
> Re-distribute or sell the resources, regardless of any modifications.
> Re-distribute or sell parts of the resources.

>In other words
>... Just do not re-distribute or sell the resources or parts of the resources as your own.

I assume Fastlane would count as redistribution, but it wouldn't necessarily be passing it off as its own? Thoughts?

I could probably send the company an email asking for clarification if needed.

**Comment by janpio** [comment]
IANAL but having them on our frameit-frames repository seems like a pretty clear redistribution to me :/ @joshdholtz Do you maybe want to contact the people of applypixels? Maybe they are fine with this.

The other option would be to add a "download those frames and create the frames for frameit locally" step. That would probably require a new action in fastlane, but that would be legally bulletproof. Toughts @mollyIV? Would the current frame generation process be something that we could adapt into an action?

**Comment by Alarson93** [comment]
Any updates?
(mostly a bump to keep this issue alive)

**Comment by nemo** [comment]
Would love it if someone could merge the above ^!

**Comment by mollyIV** [comment]
> "download those frames...

Currently that's the manual step:

https://github.com/fastlane/fastlane/blob/e3d4f17e00a607d5d36f0c7214a201267cc8025a/frameit/frames_generator/Rakefile#L31-L39

however I believe that this could be easily automated (downloading the images from a public GitHub repository).

> and create the frames for frameit locally" 

@janpio not sure I understand this part 🤔 Is there any additional action we shall perform on downloaded frames before passing them to `frameit` processing?

**Comment by janpio** [comment]
> however I believe that this could be easily automated

As this is a contributor facing action, downloading manually and checking the file content a bit is an ok thing to do.

> Is there any additional action we shall perform on downloaded frames before passing them to frameit processing?

Yes, look at what the `Rakefile` does. After you downloaded the files from Facebook Design they are "processed" to get the correct ones, do some measurements on them that is written to files (you actually implemented some of that @mollyIV!) etc.

**Comment by claesjacobsson** [comment]
Has this been solved in the latest release?

**Comment by janpio** [comment]
Partially. Fastlane now can use frames for this iPad, but they are not included in the frames by default yet as there is no clear source (licence and process) for them that could be included.

**Comment by chriswilke** [comment]
Has anyone tried to contact the provider?  I am emailing them to see we can get approval from them.  

If you have any questions, complaints, or claims with respect to the Services, you may contact us at:
Pixelresort
Danish company registration no. 32414540
Attention: Michael Flarup
Street: Egøjevej 44
Postal code: DK-4600
City: Køge
Mail: support@applypixels.com

**Comment by chriswilke** [comment]
Please note, I sent Michael the below email.

Dear Michael,

Good morning.

Ask
Can we use these iPhone frames in the frameit open source project?
https://applypixels.com/resource/ipad-pro-3rd-gen

Why?
Frameit helps automate putting frames around screenshot images for the App Store.

Frameit is part of fastlane and is licensed under the MIT License.
https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md

The images will not be sold or repackaged.
The images will be checked into the project and used by the frameit application to create screenshots that are uploaded to the app store.
Many people use frameit.

Are you ok if we proceed as long as you or your site is referenced as the provider of that content?

**Comment by mtrezza** [comment]
Can anyone please elaborate where these frame images have to be put manually to be available for `frameit`?  As I understand from the license they can be downloaded manually but not be "automatically" redistributed as part of `fastlane`, correct?

**Comment by mtrezza** [comment]
IANAL, but Apple Inc. holds the intellectual property rights to the design, so I doubt that any 3rd party license of an exact replication can be legally binding or enforceable.

Anyway, here are the steps to manually add the frames:
1. Download the frames from https://applypixels.com/resource/ipad-pro-3rd-gen
2. Modify shapes to make screen shape transparent to background
3. Export one frame in portrait as PNG with name `Apple iPad Pro (12.9-inch) (3rd generation) Space Gray` to `/Users/user/.fastlane/frameit/latest`
4. Add to `/Users/user/.fastlane/frameit/latest/offsets.json`: 
```
"iPad Pro (12.9-inch) (3rd generation)": {
  "offset": "+96+102",
  "width": 2046
}
```

**Comment by Alarson93** [comment]
@mtrezza I attempted your fix. It mostly worked, but the frame placing was very strange.

![iPad Pro (12 9-inch) (3rd generation)-01cash_bids_locations_framed](https://user-images.githubusercontent.com/16312918/61752846-ee747f80-ad72-11e9-8e5e-a947a6fdfb29.png)

In particular - the screenshot is scaled way down. But the original screen (before framing) is of normal size.
Does anyone have any thoughts as to why this may be?

**Comment by mtrezza** [comment]
@Alarson93 Looks like the frame image dimensions do not match the dimensions in `offsets.json`.

**Comment by Alarson93** [comment]
Thank you for your reply! Misunderstood what those offsets were.
I've seen made changes that accurately reflect the frame dimensions I'm currently using. The framing works correctly now.

Thank you very much!

TLDR for those that scrolled to the bottom - follow the approach @mtrezza laid out as a workaround for frameit not currently having a frame for the 3rd gen iPad Pro.

**Comment by ainame** [comment]
> Starting March 27, 2019, all new apps and app updates for iPhone or iPad, including universal apps, must be built with the iOS 12.1 SDK or later and support iPhone XS Max or the 12.9-inch iPad Pro (3rd generation). Screenshots for these devices will also be required.

**Comment by chriswilke** [comment]
@Alarson93 - Would it be possible to post the working template that you created?  I am trying to create the template using GIMP following the @mtrezza approach.

Specifically, I am not sure how to execute Step 2:
"Subtract the white screen shape from the gray device shapes to make the screen area transparent to background."

Also, I was not sure what dimensions to use for the image.  Thanks in advance for any assistance you can provide.

**Comment by Alarson93** [comment]
@chriswilke 

This is what I used for the frame provided by our designer:
`    "iPad Pro (12.9-inch) (3rd generation)": {
      "offset": "+235+252",
      "width": 4990
    }`

Dimensions of the provided frame are 5475 x 7140.
I'm pretty sure this is based on ApplyPixels (linked above).

**Comment by piuholo** [comment]
Still an issue.

**Comment by jhandguy** [comment]
I'm also needing this, let me know if I can be of any help! 👍

**Comment by Alarson93** [comment]
While it might not be a simple update, there is a process outlined above that tells you how to manually add the frame in order to get frameit working.

**Comment by Kamik423** [comment]
What I have done is using the [Apple Frames](https://www.macstories.net/ios/apple-frames-a-shortcut-for-framing-screenshots-from-every-apple-device/) Shortcut on macstories.​net. I extracted the base64 data and now have my own frames. However they do not state a license.

@viticci is it ok to include those in fastlane, or in App Store screenshots?

---

I appended

```json
    "iPad Pro (11-inch)": {
      "offset": "+120+120",
      "width": 1668
    },
    "iPad Pro (12.9-inch) (3rd generation)": {
      "offset": "+120+120",
      "width": 2046
    },
```

to `offsets.json`. The files are named

```
Apple iPad Pro (11-inch) Space Gray.png
Apple iPad Pro (12.9-inch) (3rd generation) Space Gray.png
```

It works perfectly for the 12.9 inch iPad, but not for the 11 inch one. Why is that? I tried a lot, but I cannot get it to work.

**Comment by nturpin0** [comment]
> @viticci is it ok to include those in fastlane, or in App Store screenshots?
Official apple materials were used.
https://www.macstories.net/ios/adding-device-frames-to-iphone-xs-and-xs-max-screenshots-with-shortcuts/
https://developer.apple.com/app-store/marketing/guidelines/#section-products

**Comment by Kamik423** [comment]
But he regularly uses it on his own website. And I don’t think he’d do it, if you didn’t have the right to.

**Comment by Kamik423** [comment]
<img width="678" alt="PNG image" src="https://user-images.githubusercontent.com/16278233/65390576-b680bd00-dd60-11e9-86bf-8160eee90403.png">

**Comment by nturpin0** [comment]
> But he regularly uses it on his own website. And I don’t think he’d do it, if you didn’t have the right to.

I didn't say that he didn't the right to at all, I'm not sure where that came from. I was just saying its apple press materials. Obviously they want people to use them, just download the source and read permissions there.

**Comment by Kamik423** [comment]
The [Apple marketing page this is sourced from](https://developer.apple.com/app-store/marketing/guidelines/#section-products) is specifically aimed at the App Store.

> Feature Apple product images on their own in your communications, and don’t include images of competing products. References to multiple platforms and competing products can be made only in copy or with badges.
> 
> Use Apple product images “as is” and without modification. Modifications include adding reflections, shadows, highlights, or graphic elements that appear to enter or come out of the product screen; cropping, tilting, or obstructing any part of the images; animating, flipping, or spinning the images; or creating buttons or icons with a product image. You can place promotional copy or violators beside the product images, not on top of them.

<img width="574" alt="Marketing Resources and Identity Guidelines - App Store - Apple Developer" src="https://user-images.githubusercontent.com/16278233/65391169-45440880-dd66-11e9-804e-170cbd5f9fba.png">

---

This seems fine for frameit (which does not put other devices in there) and the way the vast majority of developers use it. But some developers might still violate this license with the way they use it.

I think the right thing to do would be to put these images and and have fastlane print a license warning every time, since for most developers their usage would fit the license and this would relieve them of having to source and configure their own images.

**Comment by janpio** [comment]
That sounds like a plan @Kamik423. Are you up to creating a PR for this?

**Comment by Kamik423** [comment]
I can create the pull request. There are just two issues.

1. This would be for fastlane/frameit-frames, right? Would there have to be a second one for fastlane/fastlane for the license warning? While I could read up on ruby and try to understand the project structure of fastlane/fastlane/frameit it might make more sense for someone else to do this.
2. As stated in my [original comment on the issue](https://github.com/fastlane/fastlane/issues/14713#issuecomment-533884602): I only got it to work for the 12.9-3rd gen one. And I am not sure what I did any differently for that. I put the frames [on dropbox](https://www.dropbox.com/sh/cvitqzynxurw46x/AAB5JqvwHmSGhF7-0q7vWNX8a?dl=0), since we are allowed to use them now.
> I appended
> 
> ```json
>     "iPad Pro (11-inch)": {
>       "offset": "+120+120",
>       "width": 1668
>     },
>     "iPad Pro (12.9-inch) (3rd generation)": {
>       "offset": "+120+120",
>       "width": 2046
>     },
> ```
> 
> to `offsets.json`. The files are named
> 
> ```
> Apple iPad Pro (11-inch) Space Gray.png
> Apple iPad Pro (12.9-inch) (3rd generation) Space Gray.png
> ```
> 
> It works perfectly for the 12.9 inch iPad, but not for the 11 inch one. Why is that? I tried a lot, but I cannot get it to work.

```
[07:39:14]: Unknown device type for size iOS-iPad-11 for path './en-US/iPad Pro (11-inch)-0 main.png'
[07:39:14]: Unknown device type for size iOS-iPad-11 for path './en-US/iPad Pro (11-inch)-0 main.png'
[07:39:14]: Unknown device type for size iOS-iPad-11 for path './en-US/iPad Pro (11-inch)-0 main.png'
[07:39:14]: Couldn't find template for screenshot type 'Apple true Space Gray'
[07:39:14]: Please run `fastlane frameit download_frames` to download the latest frames
```

**Comment by Kamik423** [comment]
The issue is, that the 11 inch iPad Pro (and at least also the iPad Pro 10.5) are defined in `app_screenshots.rb`), but no file/device name is ever assigned in `screenshot.rb`:

https://github.com/fastlane/fastlane/blob/0de811cec8bc50e266b2724d96a3a3852c845e6e/deliver/lib/deliver/app_screenshot.rb#L28-L31

https://github.com/fastlane/fastlane/blob/65517d35e506a0f29f6a34cd4ee57b440ee32428/frameit/lib/frameit/screenshot.rb#L46-L51

After amending

```ruby
      when sizes::IOS_IPAD_11
        return 'iPad Pro (11-inch)'
```

in `screenshot.rb` the file generates correctly.

Is there a reason for these names missing for some devices that have simulators and thus fixed names (like `iPad Pro (11-inch)` for ... the 11 inch iPad Pro)? The 10.5 inch one is quite old now.

**Comment by janpio** [comment]
I have no idea, but most probably there is no specific reason and it just hadn't come up before.

**Comment by Kamik423** [comment]
I’d create a pull request to fastlane/frameit-frames next, I just want to confirm how I should deal with the license issue.

I’d personally just drop in a .md file referencing the different licenses on apple’s website. But nobody is going to see that.

What would the best way be?

**Comment by fire-at-will** [comment]
@Kamik423 Perhaps the additional licensing information could be printed in the `frame_downloader` script in the [print_disclaimer](https://github.com/fastlane/fastlane/blob/0c6a6d24098b9f6d7b16e4328a2048c6da4fa8df/frameit/lib/frameit/frame_downloader.rb#L54) function?

**Comment by Kamik423** [comment]
That seems like the right place, that would then be two different pull requests. One to fastlane/fastlane and one to fastlane/frameit-frames.

Another issue there is that (if I understand the frame_generator correctly) every time that script is executed it creates a new directory, which would lead to a loss of the non-facebook new frames every time it is updated.

Is there any good way to prevent that?

**Comment by patreu22** [comment]
Hey @Kamik423 , thanks for your work so far. Did you proceed in any way on this PR? I'm quite dependent on the iPad 3rd Gen support and you seem to be quite close to finalize the solution. 

Just wanted to give a little bump to this issue :)

**Comment by ainame** [comment]
Hi. I checked Facebook's site right now and found updates including iPad Pro 12.9 inch (3rd gen). They updated it on 24th October. I believe this is what people here want?
https://facebook.design/devices

If so, wouldn't all the problems be now resolved then?

**Comment by janpio** [comment]
That would indeed help a lot here!

Can you please create a new issue where you provide links to these older issues but most importantly the information you posted here? That would make it a _lot_ easier for us to get an understanding what is happening. Thanks!

**Comment by JulesMoorhouse** [comment]
> That would indeed help a lot here!
> 
> Can you please create a new issue where you provide links to these older issues but most importantly the information you posted here? That would make it a _lot_ easier for us to get an understanding what is happening. Thanks!

It looks like the device type is present but the images are missing.

**Comment by ainame** [comment]
@janpio  I created a new issue here. https://github.com/fastlane/fastlane/issues/15725
It seems that the core team needs to decide whether they keep supporting older generations's device images. Please let me know if the issue isn't clear.
